### PR TITLE
Add layout attribute support for RNN operator

### PR DESCRIPTION
**Pull request by nives-vukovic**
Since ONNX opset version 14, layout attribute has been introduced to RNN operator which allows two predefined layouts for input and output shapes.

Add corresponding reference, onnx, and verify tests.

Resolves https://github.com/migraphx-benchmark/AMDMIGraphX/issues/119

### CODE REVIEW

File: src/onnx/parse_rnn.cpp [hunk 1]
```diff
@@ -33,6 +33,29 @@ namespace migraphx {
...
 namespace onnx {
 
+void rnn_transpose_inputs(onnx_parser::node_info& info, std::vector<instruction_ref>& args)
+{
+    std::vector<int64_t> perm{1, 0, 2};
+    args[0] = info.add_instruction(make_op("transpose", {{"permutation", perm}}), args[0]);
+
+    if(args.size() == 6 and not args[5]->is_undefined())
+    {
+        args[5] = info.add_instruction(make_op("transpose", {{"permutation", perm}}), args[5]);
```
Comment (lakhinderwalia) [inline src/onnx/parse_rnn.cpp:43]: Where is this args[5] consumed, please?

File: src/onnx/parse_rnn.cpp [hunk 2]
```diff
@@ -33,6 +33,29 @@ namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 namespace onnx {
 
+void rnn_transpose_inputs(onnx_parser::node_info& info, std::vector<instruction_ref>& args)
+{
+    std::vector<int64_t> perm{1, 0, 2};
+    args[0] = info.add_instruction(make_op("transpose", {{"permutation", perm}}), args[0]);
+
+    if(args.size() == 6 and not args[5]->is_undefined())
```
Comment (lakhinderwalia) [inline src/onnx/parse_rnn.cpp:41]: args.size() is already 6, as we fill upto 6 arguments, before calling this method.

File: test/verify/test_rnn_4args_layout.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2022 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (lakhinderwalia) [inline test/verify/test_rnn_4args_layout.cpp:4]: 2015-2023, please.

File: src/onnx/parse_rnn.cpp (hunk 2 above)
Comment (nives-vukovic) [reply src/onnx/parse_rnn.cpp]: @lakhinderwalia  First check is removed.

File: src/onnx/parse_rnn.cpp (hunk 1 above)
Comment (nives-vukovic) [reply src/onnx/parse_rnn.cpp:43]: As in the https://onnx.ai/onnx/operators/onnx__RNN.html, the args[5] represent the optional argument initial_h, if the argument is not undefined it is used in calculations present in src/rewrite_rnn.cpp (rewrite_rnn::apply_vanilla_rnn)

File: src/onnx/parse_rnn.cpp [hunk 4]
```diff
@@ -33,6 +33,29 @@ namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 namespace onnx {
```
[the remark this answers is not in the document]
Comment (nives-vukovic) [reply src/onnx/parse_rnn.cpp:35]: @lakhinderwalia  Could you give an example how you would like this to look like, as we have not added the reference documentation in the parsing files before?

File: test/verify/test_rnn_4args_layout.cpp (hunk 3 above)
Comment (nives-vukovic) [reply test/verify/test_rnn_4args_layout.cpp:4]: @lakhinderwalia  Fixed.

Comment (nives-vukovic) [reply test/verify/test_rnn_4args_layout.cpp]: @lakhinderwalia Fixed.

File: src/onnx/parse_rnn.cpp (hunk 1 above)
Comment (lakhinderwalia) [reply src/onnx/parse_rnn.cpp:43]: Thanks.

**Review by lakhinderwalia — APPROVED** [review]
Happy new year :-) 
Copyrights can always be set to 2024, now or later..

**Pull request merged**